### PR TITLE
promoted group prune and sync from experimental

### PR DESCRIPTION
**Pull request by stevekuznetsov**
ref https://github.com/openshift/origin/issues/6276

This PR promotes `openshift ex {sync,prune}-groups` to `oadm groups {sync,prune}`

@deads2k PTAL

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
[test][extended:ldap_groups]

**Comment by deads2k** [comment]
Link to the doc updates please.

**Comment by deads2k** [comment]
Update your description to show where this moved to and from.

**Comment by deads2k** [comment]
Also, since we doc'ed it, maintain the location under `openshift ex` too.  Not the go package, just the commands.

**Comment by stevekuznetsov** [comment]
@deads2k doc PR made, re-instated `openshift ex {sync,prune}-groups`

**Comment by deads2k** [comment]
Add test for all four flavors in `test-cmd`.  Just a simple `openshift ex sync-groups --help` with a text check to make sure we're correctly plumbed through.

**Comment by stevekuznetsov** [comment]
Which bucket should I put that into? `help`? `admin` ?

**Comment by deads2k** [comment]
put it in help

**Comment by stevekuznetsov** [comment]
[test]

**Comment by stevekuznetsov** [comment]
again https://github.com/openshift/origin/issues/6259...
[test][extended:ldap_groups]

**Comment by stevekuznetsov** [comment]
flake https://github.com/openshift/origin/issues/6173 [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/7954/)

**Comment by stevekuznetsov** [comment]
re[test][extended:ldap_groups]

**Comment by stevekuznetsov** [comment]
@deads2k finally got green

**Comment by deads2k** [comment]
lgtm.  @liggitt any complaint on the name?

**Comment by kargakis** [comment]
Shouldn't groups prune be a part of the existing prune command?

**Comment by kargakis** [comment]
Also are these two commands what 6276 is asking for?

**Comment by stevekuznetsov** [comment]
@soltysh #6276 touches this and also another pull AFAIK
`oadm prune groups` does seem more fitting, but I really don't like the user story for `oadm prune groups` and `oadm groups sync`...

**Comment by stevekuznetsov** [comment]
@deads2k should we put `prune` under `oadm prune` or `oadm groups`? Do we have a strategy doc for this?

**Comment by deads2k** [comment]
> @deads2k should we put prune under oadm prune or oadm groups? Do we have a strategy doc for this?

You'll know it when you see it.

**Comment by deads2k** [comment]
@stevekuznetsov cross link prune to support all three calling locations.  Add  the new location to `test-cmd/help`

**Comment by stevekuznetsov** [comment]
@deads2k what should doc look like? `oadm groups prune`? `oadm prune groups`? Should we even mention it?

**Comment by deads2k** [comment]
> @deads2k what should doc look like? oadm groups prune? oadm prune groups? Should we even mention the duality?

Let's keep congruence, so `oadm groups prune`.  Don't bother mentioning aliases.

File: pkg/cmd/admin/prune/prune.go [hunk 1]
```diff
@@ -5,6 +5,7 @@ import (
 
 	"github.com/spf13/cobra"
 
+	"github.com/openshift/origin/pkg/cmd/admin/groups/sync/cli"
```
Comment (deads2k) [inline pkg/cmd/admin/prune/prune.go]: alias this package to something more reasonable so that the call below says what it means.

**Comment by stevekuznetsov** [comment]
Flake in e2e [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8032/console):
```
[INFO] Running a CLI command in a container using the service account
Waiting for pod test/cli-with-token to be running, status is Pending, pod ready: false
Error attaching, falling back to logs: error executing remote command: Error executing command in container: container not found ("cli-with-token")
I1221 15:24:52.828412       1 factory.go:103] Using in-cluster configuration
I1221 15:24:52.843709       1 factory.go:103] Using in-cluster configuration
I1221 15:24:52.845512       1 factory.go:103] Using in-cluster configuration
In project test on server https://172.30.0.1:443

2 warnings identified, use 'oc status -v' to see details.

[FAIL] !!!!! Test Failed !!!!
```

**Comment by stevekuznetsov** [comment]
@deads2k addressed package import comment

**Comment by deads2k** [comment]
lgtm, will merge on green extended.

**Comment by stevekuznetsov** [comment]
@deads2k rebased, re[test][extended:ldap_groups]

**Comment by stevekuznetsov** [comment]
@deads2k Added help-text fixes (addition of `--confirm` to examples, fixed `--type` instead of `--existing`) but *no* renames to [`sync.go`](https://github.com/openshift/origin/pull/6369/files#diff-2c2d4e7452efbd55e5b581666aea38b5). PTAL

**Comment by deads2k** [comment]
> @deads2k Added help-text fixes (addition of --confirm to examples, fixed --type instead of --existing) but no renames to sync.go. PTAL

I don't understand.  Go back in your `git reflog`, find the changes and give me a second commit so I can jump to it.

File: docs/generated/oadm_by_example_content.adoc [hunk 2]
```diff
@@ -149,6 +172,32 @@ Remove users from a group
...
+  $ oadm groups sync --sync-config=/path/to/ldap-sync-config.yaml
+
+  # Sync all groups except the ones from the blacklist file from an LDAP server
+  $ oadm groups sync --blacklist=/path/to/blacklist.txt --sync-config=/path/to/ldap-sync-config.yaml
+
+  # Sync specific groups specified in a whitelist file with an LDAP server 
+  $ oadm groups sync --whitelist=/path/to/whitelist.txt --sync-config=/path/to/sync-config.yaml
+
+  # Sync all OpenShift Groups that have been synced previously with an LDAP server
+  $ oadm groups sync --existing --sync-config=/path/to/ldap-sync-config.yaml
```
Comment (deads2k) [inline docs/generated/oadm_by_example_content.adoc]: need to regenerate

**Comment by deads2k** [comment]
> I don't understand. Go back in your git reflog, find the changes and give me a second commit so I can jump to it.

I think I understand now.  Get yourself back to green.

**Comment by stevekuznetsov** [comment]
updated docs & completions, re[test][extended:ldap_groups]

**Comment by stevekuznetsov** [comment]
e2e failing dc succeeded [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8050/console):
```
deploymentconfig "failing-dc" created
++ Retrying until success or timeout: oc get rc/failing-dc-1

[FAIL] !!!!! Test Failed !!!!
```

may be https://github.com/openshift/origin/issues/5680 ? output looks different

**Comment by stevekuznetsov** [comment]
re[test][extended:ldap_groups]

**Comment by stevekuznetsov** [comment]
flake on e2e pushing image [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8054/console):
```
[INFO] Tagging and pushing ruby-22-centos7 to 172.30.83.160:5000/cache/ruby-22-centos7:latest
...
Error from server: User "e2e-user" cannot list all buildconfigs in the cluster
```

https://github.com/openshift/origin/issues/6149

re[test][extended:ldap_groups]

**Comment by stevekuznetsov** [comment]
flake on `e2e` [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8062/console):
```
Error from server: User "e2e-user" cannot list all buildconfigs in the cluster
```

https://github.com/openshift/origin/issues/6149

**Comment by liggitt** [comment]
I think the real error is:
```
Error from server: 501: All the given peers are not reachable (failed to propose on members [https://172.18.7.165:4001] twice [last error: Unexpected HTTP status code]) [0]
!!! Error in hack/../test/end-to-end/core.sh:160
	'oc delete pod cli-with-token' exited with status 1
Call stack:
	1: hack/../test/end-to-end/core.sh:160 main(...)
Exiting with status 1
```

**Comment by stevekuznetsov** [comment]
@liggitt That seems more likely. Any opposition to me refactoring `e2e

**Comment by liggitt** [comment]
> Any opposition to me refactoring e2e to use os::cmd so we can read the output?

No strong opposition... I'd maybe start with just the ones capturing/echoing output to make the review easy, since that's the main issue that is making our lives difficult at the moment

**Comment by stevekuznetsov** [comment]
re[test][extended:ldap_groups]

**Comment by stevekuznetsov** [comment]
`TestAuthorizationSubjectAccessReview` flaked [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8076/console):
```
I1222 11:46:52.522839   20535 etcd.go:37] Deleting &etcd.Node{Key:"/kubernetes.io", Value:"", Dir:true, Expiration:(*time.Time)(nil), TTL:0, Nodes:etcd.Nodes(nil), ModifiedIndex:0x33e, CreatedIndex:0x33e} (child of &etcd.Node{Key:"", Value:"", Dir:true, Expiration:(*time.Time)(nil), TTL:0, Nodes:etcd.Nodes{(*etcd.Node)(0xc2085e8060), (*etcd.Node)(0xc2085e80c0)}, ModifiedIndex:0x0, CreatedIndex:0x0})
F1222 11:46:58.802390   20535 etcd.go:39] Unable to delete key: 100: Key not found (/kubernetes.io) [1113]
```

https://github.com/openshift/origin/issues/6065
re[test][extended:ldap_groups]

**Comment by deads2k** [comment]
[merge]

**Pull request merged**